ext/mbstring/mbstring.c

```diff
@@ -3226,43 +3226,37 @@ static int mb_recursive_encoder_detector_feed(mbfl_encoding_detector *identd, zv
 	return 0;
 } /* }}} */
 
-static int mb_recursive_convert_variable(mbfl_buffer_converter *convd, zval *var) /* {{{ */
+static bool mb_recursive_convert_variable(zval *var, const mbfl_encoding* from_encoding, const mbfl_encoding* to_encoding)
 {
-	mbfl_string string, result, *ret;
 	HashTable *ht;
 	zval *entry, *orig_var;
 
 	orig_var = var;
 	ZVAL_DEREF(var);
+
 	if (Z_TYPE_P(var) == IS_STRING) {
-		string.val = (unsigned char *)Z_STRVAL_P(var);
-		string.len = Z_STRLEN_P(var);
-		ret = mbfl_buffer_converter_feed_result(convd, &string, &result);
-		if (ret != NULL) {
-			zval_ptr_dtor(orig_var);
-			// TODO: avoid reallocation ???
-			ZVAL_STRINGL(orig_var, (char *)ret->val, ret->len);
-			efree(ret->val);
-		}
+		zend_string *ret = php_mb_convert_encoding_ex(Z_STRVAL_P(var), Z_STRLEN_P(var), to_encoding, from_encoding);
+		zval_ptr_dtor(orig_var);
+		ZVAL_STR(orig_var, ret);
```

Comment:
This is where I would like to ask someone who knows PHP internals well... am I handling the refcounts correctly?

`mb_convert_variables` replaces strings in an array or object. After converting the old `zend_string` to the new one, I am using `zval_ptr_dtor` to drop the reference to the old one, then `ZVAL_STR` to make the `zval` point to the new string.

I don't think `ZVAL_STR` modifies the refcount, and when the new `zend_string` is returned from `php_mb_convert_encoding_ex`, it should start with a refcount of 1, so hopefully we are OK here... 😕

Reply:
This seems to make sense, but you can always run some tests with Valgrind or ASAN to see if there is any issue. But been staring at this for a bit and it seems to check out

Reply:
Ah, that's a good idea.